File: leetcode/07-Dynamic-Programming/Introduction-to-Dynamic-Programming/House Robber/198. House Robber.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
    int rob(vector<int> &nums)
    {
        int l = nums.size();
        vector<int> dp(l + 2);
        for (int i = 0; i < l; ++i)
        {
            dp[i + 2] = max(dp[i] + nums[i], dp[i + 1]);
        }
        return dp[l + 1];
    }
};

/*
    看第i家的最大收益，则为偷了第i-2家和第i家的收益和偷了第i-1家的收益最大的是什么
*/

File: leetcode/07-Dynamic-Programming/DP-with-state-machine/198. House Robber.go
package dp_with_state_machine

func rob(nums []int) int {
	n := len(nums)
	dp := make([]int, n+2)
	for i := range nums {
		dp[i+2] = max(dp[i+1], dp[i]+nums[i])
	}
	return max(dp[n], dp[n+1])
}

/*
* 需要至少隔一家，那么就是最后一家或者倒数第二家
 */
